### PR TITLE
Small phase retrieval fixes, minor doc fixes, dirac combs, and faster RAM

### PR DESCRIPTION
**Pull request by tachella**
- Solves #938 #940 #941 
- Adds some efficiency fixes to `deepinv.models.RAM` which speed up approximately x2
- Adds `deepinv.utils.dirac_comb` (nice util for checking the response of spatially varying blur kernels)

### Checks to be done before submitting your PR

- [ ] `python3 -m pytest deepinv/tests` runs successfully.
- [x] `black .` runs successfully.
- [ ] `make html` runs successfully (in the `docs/` directory).
- [x] Updated docstrings related to the changes (as applicable).
- [x] Added an entry to the [changelog.rst](https://github.com/deepinv/deepinv/blob/main/docs/source/changelog.rst).

### CODE REVIEW

File: deepinv/loss/metric/distortion.py [hunk 1]
```diff
@@ -14,7 +14,7 @@
 
 
 class MAE(Metric):
-    r"""
+    r"""deepinv.metric.MAE(complex_abs, reduction, norm_inputs, center_crop, \*\*kwargs)
```
Comment (mh-nguyen712) [inline deepinv/loss/metric/distortion.py]: Why adding this? To give hint on how to use the function? If yes suggest adding :class:`...` and a new line after this.

Comment (mh-nguyen712) [reply deepinv/loss/metric/distortion.py]:
```suggestion
    r""":class:`deepinv.metric.MAE(complex_abs, reduction, norm_inputs, center_crop, \*\*kwargs)`
    
```

File: deepinv/loss/metric/distortion.py [hunk 2]
```diff
@@ -52,7 +52,7 @@ def metric(self, x_net, x, *args, **kwargs):
 
 
 class MSE(Metric):
-    r"""
+    r"""deepinv.metric.MSE(complex_abs, reduction, norm_inputs, center_crop, \*\*kwargs)
```
Comment (mh-nguyen712) [inline deepinv/loss/metric/distortion.py]: Same here

Comment (mh-nguyen712) [reply deepinv/loss/metric/distortion.py]: image
add a new line

File: deepinv/loss/metric/distortion.py [hunk 3]
```diff
@@ -522,7 +522,7 @@ def metric(
 
 
 class SpectralAngleMapper(Metric):
-    r"""
+    r"""deepinv.metric.SpectralAngleMapper(train_loss, reduction, norm_inputs, center_crop, \*\*kwargs)
```
Comment (mh-nguyen712) [inline deepinv/loss/metric/distortion.py]: same here

File: deepinv/models/ram.py [hunk 4]
```diff
@@ -188,7 +187,9 @@ def realign_input(self, x, physics, y, sigma):
         gamma = gamma[(...,) + (None,) * (x.dim() - 1)]
         gamma = gamma * self.fact_realign
         gamma = gamma.clamp(min=1e-8)  # clamp to avoid negative or zero gamma
-        model_input = physics.prox_l2(x, y, gamma=gamma)
+        model_input = physics.base.prox_l2(
```
Comment (mh-nguyen712) [inline deepinv/models/ram.py]: I think it's safer to just call `physics.prox_l2` here. Since if it's a DecomposablePhysics, `physics.prox_l2` should use the fast implementation. If it's not (I think that it's not the case), I would change the behavior in physics class, not here.

File: deepinv/models/ram.py [hunk 5]
```diff
@@ -437,18 +476,18 @@ def krylov_embeddings(y, p, factor, v=None, N=4, x_init=None):
     """
 
     if x_init is None:
-        x = p.A_adjoint(y)
+        x = y
     else:
-        x = x_init.clone()
+        x = x_init  # .clone()
```
Comment (mh-nguyen712) [inline deepinv/models/ram.py]:
```suggestion
        x = x_init
```

File: deepinv/models/ram.py [hunk 6]
```diff
@@ -437,18 +476,18 @@ def krylov_embeddings(y, p, factor, v=None, N=4, x_init=None):
...
-    AtA = lambda u: p.A_adjoint(p.A(u)) * norm  # Define the linear operator
+    AtA = lambda u: p.A_adjoint_A(u) * norm  # Define the linear operator
 
     v = v if v is not None else torch.zeros_like(x)
 
-    out = x.clone()
+    out = x  # .clone()
     # Compute Krylov basis
-    x_k = x.clone()
+    x_k = x  # .clone()
```
Comment (mh-nguyen712) [inline deepinv/models/ram.py]: Just a remark that It could be unsafe if AtA has in-place operation.

File: deepinv/sampling/diffusion.py [hunk 7]
```diff
@@ -568,7 +569,7 @@ class DPS(Reconstructor):
     def __init__(
         self,
         model,
-        data_fidelity,
+        data_fidelity=L2(),
```
Comment (mh-nguyen712) [inline deepinv/sampling/diffusion.py]: I think it's safer to use: 
```suggestion
        data_fidelity=None,
```
then 
``` if data_fidelity is None:
        data_fidelity = L2()
```
This can prevent bugs, since default arguments are evaluated only once, at definition time

File: deepinv/loss/metric/distortion.py (hunk 1 above)
Comment (tachella) [reply deepinv/loss/metric/distortion.py]: @mh-nguyen712 It is for overwriting the way the function signature is displayed in the docs (see #938 for an example). The issue is that these classes do not have custom `__init__`, so they inherit the init from the parent `deepinv.metric.Metric` which has an undesired input `metric` (this is already defined as the MAE in this case). 

In order to overwrite, we need to add the string at the very first line with a new description of input params as I did in the PR.

File: deepinv/loss/metric/distortion.py (hunk 2 above)
Comment (tachella) [reply deepinv/loss/metric/distortion.py]: (see my comment above)

File: deepinv/loss/metric/distortion.py (hunk 3 above)
Comment (tachella) [reply deepinv/loss/metric/distortion.py]: same, see comment above

File: deepinv/models/ram.py (hunk 4 above)
Comment (tachella) [reply deepinv/models/ram.py]: agreed, I move this to the parent class `LinearPhysicsMultiScaler`, which now uses the base class `A_dagger` and `prox_l2` if it is run at `scale=0` (full res)

File: deepinv/models/ram.py (hunk 5 above)
Comment (tachella) [reply deepinv/models/ram.py]: done

File: deepinv/models/ram.py (hunk 6 above)
Comment (tachella) [reply deepinv/models/ram.py]: I kept it just in case

File: deepinv/sampling/diffusion.py (hunk 7 above)
Comment (tachella) [reply deepinv/sampling/diffusion.py]: done

File: deepinv/loss/metric/distortion.py (hunk 1 above)
Comment (mh-nguyen712) [reply deepinv/loss/metric/distortion.py]: Thanks for the clarification. Can you just add a newline after the signature if possible, because when using type hint (eg in vscode) it gives the docstring right after the signature (see photo in my previous comment).
(Not really important, you can ignore it if you feel it's not necessary)

File: deepinv/models/ram.py (hunk 4 above)
Comment (mh-nguyen712) [reply deepinv/models/ram.py]:
```suggestion
        model_input = physics.prox_l2(
```

File: deepinv/physics/wrappers.py [hunk 8]
```diff
@@ -129,6 +129,54 @@ def A_adjoint(self, y, scale=None, **kwargs):
...
+        .. math::
+
+            \underset{x}{\arg\min} \; \frac{\gamma}{2}\|Ax-y\|^2 + \frac{1}{2}\|x-z\|^2
+
+        If the scale is set to 0, it uses the base physics proximal operator, which might have a more efficient implementation.
+
+        :param torch.Tensor y: measurements tensor
+        :param torch.Tensor z: signal tensor
+        :param float gamma: hyperparameter of the proximal operator
+        :param str solver: solver to use for the proximal operator, see :meth:`deepinv.utils.least_squares` for details
```
Comment (mh-nguyen712) [inline deepinv/physics/wrappers.py]:
```suggestion
        :param str solver: solver to use for the proximal operator, see :func:`deepinv.utils.least_squares` for details
```

File: deepinv/physics/wrappers.py [hunk 9]
```diff
@@ -129,6 +129,78 @@ def A_adjoint(self, y, scale=None, **kwargs):
...
+        .. math::
+
+            \underset{x}{\arg\min} \; \frac{\gamma}{2}\|Ax-y\|^2 + \frac{1}{2}\|x-z\|^2
+
+        If the scale is set to 0, it uses the base physics proximal operator, which might have a more efficient implementation.
+
+        :param torch.Tensor y: measurements tensor
+        :param torch.Tensor z: signal tensor
+        :param float gamma: hyperparameter of the proximal operator
+        :param str solver: solver to use for the proximal operator, see :func:`deepinv.utils.least_squares` for details
```
Comment (mh-nguyen712) [inline deepinv/physics/wrappers.py]:
```suggestion
        :param str solver: solver to use for the proximal operator, see :func:`deepinv.optim.utils.least_squares` for details
```

File: deepinv/physics/forward.py [hunk 10]
```diff
@@ -706,6 +706,11 @@ def prox_l2(
         :param torch.Tensor y: measurements tensor
         :param torch.Tensor z: signal tensor
         :param float gamma: hyperparameter of the proximal operator
+        :param str solver: solver to use for the proximal operator, see :func:`deepinv.utils.least_squares` for details
```
Comment (mh-nguyen712) [inline deepinv/physics/forward.py]:
```suggestion
        :param str solver: solver to use for the proximal operator, see :func:`deepinv.optim.utils.least_squares` for details
```

**Review by mh-nguyen712 — CHANGES_REQUESTED** [review]
LGTM!
I only have to comments related to function ref to make the docs pass.

**Comment by tachella** [comment]
all tests seem to pass now, @mh-nguyen712 can we merge?
